gh-115999: Make list and tuple iteration more thread-safe.

Yhg1s opened the pull request:
Make tuple iteration more thread-safe, and actually test concurrent iteration of tuple, range and list. (This is prep work for enabling specialization of FOR_ITER in free-threaded builds.) The basic premise is:

 - Iterating over a shared _iterable_ (list, tuple or range) should be safe, not involve data races, and behave like iteration normally does.

 - Using a shared _iterator_ should not crash or involve data races, and should only produce items regular iteration would produce. It is _not_ guaranteed to produce all items, or produce each item only once. (This is not the case for range iteration even after this PR.)

Providing stronger guarantees is possible for some of these iterators, but it's not always straight-forward and can significantly hamper the common case. Since iterators in general aren't shared between threads, and it's simply impossible to concurrently use many iterators (like generators), better to make sharing iterators without explicit synchronization clearly wrong.

Specific issues fixed in order to make the tests pass:

 - List iteration could occasionally fail an assertion when a shared list was shrunk and an item past the new end was retrieved concurrently. There's still some unsafety when deleting/inserting multiple items through for example slice assignment, which uses memmove/memcpy.

 - Tuple iteration could occasionally crash when the iterator's reference to the tuple was cleared on exhaustion. Like with list iteration, in free-threaded builds we can't safely and efficiently clear the iterator's reference to the iterable (doing it safely would mean extra, slow refcount operations), so just keep the iterable reference around.

<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

Yhg1s committed:
Make list, tuple and range iteration more thread-safe, and actually test
concurrent iteration. (This is prep work for enabling specialization of
FOR_ITER in free-threaded builds.) The basic premise is:

 - Iterating over a shared _iterable_ (list, tuple or range) should be safe,
   not involve data races, and behave like iteration normally does.

 - Using a shared _iterator_ should not crash or involve data races, and
   should only produce items regular iteration would produce. It is _not_
   guaranteed to produce all items, or produce each item only once.

Providing stronger guarantees is possible for some of these iterators, but
it's not always straight-forward and can significantly hamper the common
case. Since iterators in general aren't shared between threads, and it's
simply impossible to concurrently use many iterators (like generators),
better to make sharing iterators without explicit synchronization clearly
wrong.

Specific issues fixed in order to make the tests pass:

 - List iteration could occasionally crash when a shared list wasn't already
   marked as shared when reallocated.

 - Tuple iteration could occasionally crash when the iterator's reference to
   the tuple was cleared on exhaustion. Like with list iteration, in
   free-threaded builds we can't safely and efficiently clear the iterator's
   reference to the iterable (doing it safely would mean extra, slow
   refcount operations), so just keep the iterable reference around.

 - Fast range iterators (for integers that fit in C longs) shared between
   threads would sometimes produce values past the end of the range, because
   the iterators use two bits of state that we can't efficiently update
   atomically. Rewriting the iterators to have a single bit of state is
   possible, but probably means more math for each iteration and may not be
   worth it.

 - Long range iterators (for other numbers) shared between threads would
   crash catastrophically in a variety of ways. This now uses a critical
   section. Rewriting this to be more efficient is probably possible, but
   since it deals with arbitrary Python objects it's difficult to get right.

There seem to be no more exising races in list_get_item_ref, so drop it from
the tsan suppression list.

Yhg1s committed:
Fix lint issues.

corona10 commented on Objects/rangeobject.c:
```diff
@@ -979,75 +998,87 @@ typedef struct {
-    Py_INCREF(r->len);
-    return r->len;
+    PyObject *len;
+    Py_BEGIN_CRITICAL_SECTION(r);
```
Out of curiosity, using atomic operation is not enough for this operation?
(Maybe with _Py_TryIncref?)

Yhg1s replied:
It's not enough to make this an atomic read without also making the writes in the other critical sections atomic (in addition to using TryIncrefCompare here); it's still a data race to write non-atomically even if all reads are atomic. I didn't really want to do all that work just to avoid a critical section in this function, when this isn't a particularly common or performance-sensitive function, and an uncontested critical section is probably just as fast anyway.

Yhg1s committed:
Restore the original guard for list_get_item_ref's fast path, since it's
actually correct and the real problem was an incorrect assert. The fast path
still contains notionally unsafe uses of memcpy/memmove, so add
list_get_item_ref back to the TSan suppressions file.

Yhg1s committed:
Merge branch 'main' into list-realloc

Yhg1s committed:
Fix test failures in test_sys because of the changed size of range
iterators, and fix build failures because labels can't technically be at the
end of compound statements (a statement must follow the label, even if it's
empty).

Yhg1s committed:
Merge branch 'list-realloc' of github.com:Yhg1s/cpython into list-realloc

mpage commented on Lib/test/test_free_threading/test_iteration.py:
```diff
@@ -0,0 +1,123 @@
+        """Test iteration over a shared container"""
+        seq = self.make_testdata(NUMITEMS)
+        results = []
+        start = threading.Event()
```
If you are trying to synchronize when the workers start running I think `threading.Barrier` more closely matches that behavior.

Same comment applies below.

colesbury commented on Objects/listobject.c:
```diff
@@ -937,7 +937,7 @@ list_ass_slice_lock_held(PyListObject *a, Py_ssize_t ilow, Py_ssize_t ihigh, PyO
     for (k = 0; k < n; k++, ilow++) {
         PyObject *w = vitem[k];
-        item[ilow] = Py_XNewRef(w);
+        FT_ATOMIC_STORE_PTR_RELAXED(item[ilow], Py_XNewRef(w));
```
I think this should be `FT_ATOMIC_STORE_PTR_RELEASE`. We generally need `release` when writing pointers that may be load and dereferenced outside of the lock to ensure that the previously written contents are visible before the pointer itself.

colesbury commented on Objects/rangeobject.c:
```diff
@@ -979,75 +998,89 @@ typedef struct {
 }
 
 static PyObject *
 longrangeiter_setstate(longrangeiterobject *r, PyObject *state)
```
If we need to wrap functions in critical sections, I think we should generally convert them to argument clinic, if possible, and use the `@critical_section` annotation to keep the boilerplate mostly in a generated file.

That's not yet possible for `longrangeiter_next` (because it's a tp_iternext, not a PyMethodDef entry), but it's doable for the other functions.

Yhg1s replied:
I've dropped the range iterator fixes (fast and long) from this PR, and I'll leave it for a separate change. (I'll file an issue.)

colesbury commented on Objects/rangeobject.c:
```diff
@@ -966,6 +982,9 @@ fast_range_iter(long start, long stop, long step, long len)
     it->len = len;
+#ifdef Py_GIL_DISABLED
+    it->stop = stop;
+#endif
```
I'm not enthusiastic about the idea of growing the iterator size. Can we unconditionally replace `len` with `stop`?

Yhg1s committed:
Fix tupleiter_reduce for test_iter's assumptions, and make tupleiter_len and
tupleiter_setstate appropriately relaxed.

Yhg1s committed:
Undo the range iterator changes (leave them for later), adjust the range
iterator test to accept the current behaviour of the fast range iterator and
avoid testing the (very unsafe) long range iterator, use threading.Barrier
where appropriate, and add comments to the unsafe uses of memcpy/memmove in
free-threaded builds.

Yhg1s committed:
Remove edit snafu.

Yhg1s committed:
Merge branch 'main' into list-realloc

eendebakpt commented on Objects/tupleobject.c:
```diff
@@ -1020,14 +1020,17 @@ tupleiter_next(PyObject *self)
         return Py_NewRef(item);
     }
 
+#ifndef Py_GIL_DISABLED
```
Since we do not set `it_seq` to NULL in the free-threading build we can change the lines
``` 
if (seq == NULL)
        return NULL;
```
into
```
#ifdef Py_GIL_DISABLED
assert(seq != NULL);
#else
if (seq == NULL)
        return NULL;
#endif
```

eendebakpt commented on Objects/tupleobject.c:
```diff
@@ -1036,8 +1039,14 @@ tupleiter_len(PyObject *self, PyObject *Py_UNUSED(ignored))
     Py_ssize_t len = 0;
+#ifdef Py_GIL_DISABLED
+    Py_ssize_t idx = FT_ATOMIC_LOAD_SSIZE_RELAXED(it->it_index);
+    if (it->it_seq && idx >= 0 && idx < PyTuple_GET_SIZE(it->it_seq))
```
```suggestion
    if (idx < PyTuple_GET_SIZE(it->it_seq))
```
?

The  `it->it_seq` is not set to null any longer and idx cannot be negative. (I do not think `it_index` can overflow through the increments in `tupleiter_next`)

Yhg1s committed:
Extend tests a little, add suppressions with links to GH issues.

Yhg1s committed:
Address reviewer comments.

Yhg1s committed:
Turn the last few stores/loads of list iterator/reversed iterator into
relaxed atomics. (These are not performance-sensitive parts of the code, and
this lets us detect races elsewhere involving the iterators.)

Yhg1s committed:
Merge branch 'main' into list-realloc

mpage commented on Lib/test/test_free_threading/test_iteration.py:
```diff
@@ -0,0 +1,132 @@
+# Long range iterators are not thread-safe yet.
+# class ContendedLongRangeIterationTest(ContendedTupleIterationTest):
+#     def make_testdata(self, n):
+#         return range(0, sys.maxsize*n, sys.maxsize)
```
I think we can delete this and add it in the PR that makes the iterators thread-safe.

```suggestion
```

Yhg1s replied:
Done.

Yhg1s committed:
Address reviewer comments.

Yhg1s committed:
Merge branch 'list-realloc' of github.com:Yhg1s/cpython into list-realloc

Yhg1s committed:
Merge branch 'main' into list-realloc

colesbury reviewed (approved):
LGTM
